src/engine/controls/macrocontrol.cpp

```diff
@@ -0,0 +1,336 @@
+#include "macrocontrol.h"
+
+#include <QRegExp>
+
+#include "moc_macrocontrol.cpp"
+#include "track/track.h"
+
+namespace {
+constexpr uint kRecordingTimerInterval = 100;
+constexpr size_t kRecordingQueueSize = kRecordingTimerInterval / 10;
+
+const QString kRecordingSuffix = QStringLiteral(" [Recording]");
+const QString kInterruptedSuffix = QStringLiteral(" [Interrupted]");
+const QRegExp kAutoGeneratedLabelRegExp(QStringLiteral("[0-9. ]+(\\[\\w*\\])?"));
+} // namespace
+
+ConfigKey MacroControl::getConfigKey(const QString& name) {
+    return ConfigKey(m_group, QStringLiteral("macro_%1_%2").arg(m_slot).arg(name));
+}
+
+MacroControl::MacroControl(const QString& group, UserSettingsPointer pConfig, int slot)
+        : EngineControl(group, pConfig),
+          m_slot(slot),
+          m_queuedJumpTarget(-1),
+          m_recordedActions(kRecordingQueueSize),
+          m_nextActionIndex(0),
+          m_COStatus([this]() { return getConfigKey("status"); }()),
+          m_CORecord([this]() { return getConfigKey("record"); }()),
+          m_COPlay([this]() { return getConfigKey("play"); }()),
+          m_COEnable([this]() { return getConfigKey("enable"); }()),
+          m_COLoop([this]() { return getConfigKey("loop"); }()),
+          m_activate([this]() { return getConfigKey("activate"); }()),
+          m_toggle([this]() { return getConfigKey("toggle"); }()),
+          m_clear([this]() { return getConfigKey("clear"); }()) {
```

Comment:
this is dubious because it relies on the class already being partially initialized. I think the better option would be to just pass in the `group` along with `getConfigKey` and make it `static`...